server/routerlicious/packages/services-utils/src/tokenManager.ts

```diff
@@ -0,0 +1,109 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import {
+	IWebSocket,
+	IWebSocketTracker,
+	IJsonWebTokenManager,
+} from "@fluidframework/server-services-core";
+import { Lumberjack } from "@fluidframework/server-services-telemetry";
+import { NetworkError } from "@fluidframework/server-services-client";
+
+export class WebSocketTracker implements IWebSocketTracker {
+	// Map of socket id to socket object
+	private readonly socketIdToSocketMap: Map<string, IWebSocket>;
+	// Map of composite token id to socket ids. It assumes one token could be used for multiple sockets
+	private readonly tokenIdToSocketIdMap: Map<string, Set<string>>;
+	// Map of socketId to token ids. It assumes one socket could be used for connections with multiple tokens
+	private readonly socketIdToTokenIdMap: Map<string, Set<string>>;
+
+	constructor() {
+		this.socketIdToSocketMap = new Map();
+		this.tokenIdToSocketIdMap = new Map();
+		this.socketIdToTokenIdMap = new Map();
+	}
+
+	public addSocket(compositeTokenId: string, webSocket: IWebSocket) {
+		if (this.tokenIdToSocketIdMap.has(compositeTokenId)) {
+			this.tokenIdToSocketIdMap.get(compositeTokenId)?.add(webSocket.id);
```

Comment:
>this.tokenIdToSocketIdMap.get(compositeTokenId)? [](http://example.com/codeflow?start=3&length=48)

Could this be null? and what if it is null?

Reply:
The has function should already check if Map has the key or not and Map returns undefined if no key found (which should be already checked by has function). The question mark here is to silence the typescript compiler since it thinks get may return undefined. Maybe a exclamation mark(!) is more appropriate here? I think null and undefined are same in this case and we always insert a set as the value (the set function). As long as Map has the key, it cannot be null. Or maybe I misunderstand your comments?

Reply:
I thought `tokenIdToSocketIdMap` could have the `compositeTokenId` key with a null value because of the question mark. If the value can't be null, an exclamation mark should a more suitable choice.

Reply:
I refactor it a little bit. No more question marks needed